abci/client/local_client.go

```diff
@@ -76,8 +78,8 @@ func (app *localClient) InfoAsync(ctx context.Context, req types.RequestInfo) (*
 }
 
 func (app *localClient) DeliverTxAsync(ctx context.Context, params types.RequestDeliverTx) (*ReqRes, error) {
-	app.mtx.Lock()
-	defer app.mtx.Unlock()
+	app.mtx.RLock()
```

Comment:
this must be a write-lock.

Reply:
does this mean that application implementations are depending on being called within a lock that they don't manage? 

In any case, I've made these all use the write lock.